File: CmdMsg.java
import java.io.Serializable;

/**
* Object to contain command and messages
* @author Alex Lewtschuk, Kai Sorensen
*/
public class CmdMsg extends Object implements Serializable {
    String cmd;
    String msg;
    String event;

    /**
     * Constructor 
     * @param command
     * @param messsage
     */
    public CmdMsg(String command, String messsage){
        this.cmd = command;
        this.msg = messsage;
    }

    // /**
    //  * Constructor for an event if we needed it 
    //  * @param event
    //  */
    // public CmdMsg(String event){
    //     this.msg = event;
    // }


    /**
     * This updates the objects contents
     * @param updateCmd
     * @param updateMsg
     */
    public void updateContents(String updateCmd, String updateMsg){
        this.cmd = updateCmd;
        this.msg = updateMsg;
    }

    /**
     * Alows modification of nickname
     * @param trimmedNick
     */
    public void updateNickname(String trimmedNick){
        this.msg = trimmedNick;
    }

    /**
     * Gets the command part of the object
     * @return
     */
    public String getCommand(){
        return cmd;
    }

    /**
     * Gets the message part of the object
     * @return
     */
    public String getMessage(){
        return msg;
    }

    /**
     * Gets the possible event in the object
     * @return
     */
    public String getEvent(){
        return msg;
    }

    @Override
    public String toString(){
        return "Object contains command: " + cmd + " argument: " + msg;
    }

    /**
     * Checks if the nickncame in the object has invalid characters 
     * @param msg
     * @return
     */
    public boolean checkBadNickChars(String msg) {
        final Character[] specialChars = {'\t', '\n', '/', '\\'};
        if(msg == null || msg.isEmpty()){
            return true;
        }
        for(int i = 0; i < specialChars.length; i++){
            for(int j = 0; j < msg.length(); j++){
                char c = msg.charAt(j);
                if(c == specialChars[i]){
                    return true;
                }
            }
        }
        return false;
    }

    /**
     * Checks if the first character is blank. Used as an edge case check
     * @param msg
     * @return
     */
    public boolean firstNickCharIsSpace(String msg){
        if(msg.startsWith(" ")){
            System.out.println("The first character is a space. Please reenter your name\n");
            return true;
        }
        return false;
    }
}

File: Go/app/cmdmsg/cmdmsg.go
package cmdmsg

import (
	"fmt"
	"strings"
)

/*
Defines the struct for the command message that will be sent accross the network
*/
type CmdMsg struct {
	Cmd string `json:"cmd"`
	Msg string `jason:"msg"`
}

// Updates the struct's contents
func UpdateContents(cmdMsg *CmdMsg, updateCmd string, updateMsg string) {
	cmdMsg.Cmd = updateCmd
	cmdMsg.Msg = updateMsg
}

// Allows modification of the nickname
func UpdateNickname(cmdMsg *CmdMsg, trimmedNick string) {
	cmdMsg.Msg = trimmedNick
}

// func GetCommand(cmdMsg *CmdMsg) string {
// 	cmd := cmdMsg.Cmd
// 	return cmd
// }

// func GetMessage(cmdMsg *CmdMsg) string {
// 	msg := cmdMsg.Msg
// 	return msg
// }

func ToString(cmdmsg *CmdMsg) string {
	return "CmdMsg contains command " + cmdmsg.Cmd + " and argument: " + cmdmsg.Msg
}

func CheckBadNickChars(msg string) bool {

	if strings.HasPrefix(msg, " ") {
		fmt.Println("The first character is a space. Please reenter your name")
		return true
	}

	forbiddenChars := []rune{'\t', '\n', '/', '\\'}
	if msg == "" {
		return true
	}

	for i := 0; i < len(msg); i++ {
		for i2 := 0; i2 < len(forbiddenChars); i++ {
			runeAt := strings.IndexRune(msg, forbiddenChars[i2])
			if runeAt != -1 {
				return true
			}
		}
	}
	return false
}
